includes/class-wc-subscriptions-data-copier.php

```diff
@@ -0,0 +1,461 @@
+		'_order_key',
+		'_edit_lock',
+		'_wc_points_earned',
+		'_transaction_id',
+		'_billing_interval',
+		'_billing_period',
+		'_subscription_resubscribe',
+		'_subscription_renewal',
+		'_subscription_switch',
+		'_payment_method',
+		'_payment_method_title',
+		'_suspension_count',
+		'_requires_manual_renewal',
+		'_cancelled_email_sent',
+		'_trial_period',
+		'_created_via',
+		'_order_stock_reduced',
+	];
+
+	/**
+	 * The subscription or order being copied.
+	 *
+	 * @var WC_Order
+	 */
+	private $from_object = null;
+
+	/**
+	 * The subscription or order being copied to.
+	 *
+	 * @var WC_Order
+	 */
+	private $to_object = null;
+
+	/**
+	 * The type of copy. Can be 'subscription' or 'renewal'.
+	 *
+	 * Used in dynamic filters to allow third parties to target specific meta keys in different copying contexts.
+	 *
+	 * @var string
+	 */
+	private $copy_type = '';
+
+	/**
+	 * Copies data from one object to another.
+	 *
+	 * This function acts as a publicly accessible wrapper for obtaining an instance of the copier and completing the copy.
+	 *
+	 * @param WC_Order $from_object The object to copy data from.
+	 * @param WC_Order $to_object   The object to copy data to.
+	 * @param string   $copy_type   Optional. The type of copy. Can be 'subscription', 'parent', 'renewal_order' or 'resubscribe_order'. Default is 'subscription'.
+	 */
+	public static function copy( $from_object, $to_object, $copy_type = self::DEFAULT_COPY_TYPE ) {
+		$instance = new self( $from_object, $to_object, $copy_type );
+		$instance->copy_data();
+	}
+
+	/**
+	 * Constructor.
+	 *
+	 * @param WC_Order $from_object The object to copy data from.
+	 * @param WC_Order $to_object   The object to copy data to.
+	 * @param string   $copy_type   Optional. The type of copy. Can be 'subscription', 'parent', 'renewal_order' or 'resubscribe_order'. Default is 'subscription'.
+	 */
+	public function __construct( $from_object, $to_object, $copy_type = self::DEFAULT_COPY_TYPE ) {
+		$this->from_object = $from_object;
+		$this->to_object   = $to_object;
+		$this->copy_type   = $copy_type;
+	}
+
+	/**
+	 * Copies the data from the "from" object to the "to" object.
+	 */
+	public function copy_data() {
+
+		if ( $this->is_using_cpt_data_stores() ) {
+			$data_array = $GLOBALS['wpdb']->get_results( $this->get_deprecated_meta_query(), ARRAY_A );
+			$data       = wp_list_pluck( $data_array, 'meta_value', 'meta_key' );
+		} else {
+			$data  = $this->get_meta_data();
+			$data += $this->get_order_data();
+			$data += $this->get_operational_data();
+			$data += $this->get_address_data();
+
+			// Remove any excluded meta keys.
+			$data = $this->filter_excluded_meta_keys_via_query( $data );
+		}
+
+		$data = $this->apply_deprecated_filter( $data );
+		$data = apply_filters( "wc_subscriptions_{$this->copy_type}_data", $data, $this->to_object, $this->from_object );
+		$data = apply_filters( 'wc_subscriptions_object_data', $data, $this->to_object, $this->from_object, $this->copy_type );
```

Comment:
I like all the careful discussion and consideration in this PR + reviews to get these new hooks right (naming, design, intention etc).

I'm thinking we should start documenting our hooks consistently. The idea is that we have a single source of truth for each hook – why it's there, how it's intended to be used, when it was introduced - and of course we can also document limitations in the same place (e.g. not recommended, deprecated etc). 

Here are some options for how we might do this:

- Docblock for each hook, [similar to Woo core](https://github.com/woocommerce/woocommerce/blob/bf61c6bf192f95eba44a58741256db3efbd443b1/plugins/woocommerce/includes/emails/class-wc-email.php#L329-L337). We could optionally run a tool (PHPdoc?) to generate documentation for hooks (bonus points!).
- Manually document in a google sheet or markdown file.
- Something else :)

What do y'all think about documenting new/changed hooks as we go?

Reply:
I've added filter comments in [178d850](https://github.com/Automattic/woocommerce-subscriptions-core/pull/229/commits/178d8505a8bd0c62eedbab4887f5b1def7e08e0e)

Reply:
Thanks James!